service/internal/checksum/middleware_setup_context.go

```diff
@@ -31,13 +40,17 @@ func (m *setupInputContext) HandleInitialize(
 ) (
 	out middleware.InitializeOutput, metadata middleware.Metadata, err error,
 ) {
-	// Check if validation algorithm is specified.
+	var algorithm string
+	var ok bool
 	if m.GetAlgorithm != nil {
-		// check is input resource has a checksum algorithm
-		algorithm, ok := m.GetAlgorithm(in.Parameters)
-		if ok && len(algorithm) != 0 {
-			ctx = internalcontext.SetChecksumInputAlgorithm(ctx, algorithm)
+		algorithm, ok = m.GetAlgorithm(in.Parameters)
+	}
+
+	if m.RequireChecksum || m.RequestChecksumCalculation == aws.RequestChecksumCalculationWhenSupported || ok {
```

Comment:
This works but the double-if is a little uncomfortable. Suggested workflow to improve the ergonomics:

```go
if alg, ok := m.GetAlgorithm(in.Parameters); ok {
    ctx = ... // user set an algorithm, nothing else matters
    return next.HandleInitialize(ctx, in)
}

if m.RequireChecksum || m.RequestChecksumCalculation == aws.RequestChecksumCalculationWhenSupported {
    ctx = ... // default to crc32
}
return next.HandleInitialize(ctx, in)